drivers/cache/cache_aspeed.c

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright (c) 2022 ASPEED Technology Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <arch/arm/aarch32/cortex_m/cmsis.h>
+#include <zephyr/drivers/syscon.h>
+
+/*
+ * cache area control: each bit controls 32KB cache area
+ *	1: cacheable
+ *	0: no-cache
+ *
+ *	bit[0]: 1st 32KB from 0x0000_0000 to 0x0000_7fff
+ *	bit[1]: 2nd 32KB from 0x0000_8000 to 0x0000_ffff
+ *	...
+ *	bit[22]: 23th 32KB from 0x000a_8000 to 0x000a_ffff
+ *	bit[23]: 24th 32KB from 0x000b_0000 to 0x000b_ffff
+ */
+#define CACHE_AREA_CTRL_REG	0xa50
+#define CACHE_INVALID_REG	0xa54
+#define CACHE_FUNC_CTRL_REG	0xa58
+
+#define CACHED_SRAM_ADDR	CONFIG_SRAM_BASE_ADDRESS
+#define CACHED_SRAM_SIZE	KB(CONFIG_SRAM_SIZE)
+#define CACHED_SRAM_END		(CACHED_SRAM_ADDR + CACHED_SRAM_SIZE - 1)
+
+#define CACHE_AREA_SIZE_LOG2	15
+#define CACHE_AREA_SIZE		(1 << CACHE_AREA_SIZE_LOG2)
+
+#define DCACHE_INVALID(addr)	(BIT(31) | ((addr & GENMASK(10, 0)) << 16))
+#define ICACHE_INVALID(addr)	(BIT(15) | ((addr & GENMASK(10, 0)) << 0))
+
+#define ICACHE_CLEAN		BIT(2)
+#define DCACHE_CLEAN		BIT(1)
+#define CACHE_EANABLE		BIT(0)
+
+/* cache size = 32B * 128 = 4KB */
+#define CACHE_LINE_SIZE_LOG2	5
+#define CACHE_LINE_SIZE		(1 << CACHE_LINE_SIZE_LOG2)
+#define N_CACHE_LINE		128
+#define CACHE_ALIGNED_ADDR(addr) \
+	((addr >> CACHE_LINE_SIZE_LOG2) << CACHE_LINE_SIZE_LOG2)
+
+/* prefetch buffer */
+#define PREFETCH_BUF_SIZE	CACHE_LINE_SIZE
+
+static void aspeed_cache_init(void)
+{
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(syscon));
+	uint32_t start_bit, end_bit;
+
+	/* set all cache areas to no-cache by default */
+	syscon_write_reg(dev, CACHE_FUNC_CTRL_REG, 0);
+
+	/* calculate how many areas need to be set */
+	start_bit = CACHED_SRAM_ADDR >> CACHE_AREA_SIZE_LOG2;
+	end_bit = CACHED_SRAM_END >> CACHE_AREA_SIZE_LOG2;
+	syscon_write_reg(dev, CACHE_AREA_CTRL_REG, GENMASK(end_bit, start_bit));
```

Comment:
Is it possible that the `end_bit `is too large that causing the `uint32_t ` overflow? does it matter in your system?

Reply:
Thank you for your comment.  The max RAM size in this SOC is 768KB so theoretically it wouldn't overflow.  But I guess I can add a ceil for ```start_bit``` and ```end_bit```.
 
 ````
 start_bit = MIN(31, CACHED_SRAM_ADDR >> CACHE_AREA_SIZE_LOG2);
 end_bit = MIN(31, CACHED_SRAM_END >> CACHE_AREA_SIZE_LOG2);
 ```

Reply:
Got it. it's up to you, as long as it works for your system :)

Reply:
Re-pushed.  Thanks for your review.